Search: Add metric for position of first click

jtibshirani opened the pull request:
This change adds a prometheus metric for the position of the user's first search
result click. Currently we just track the position for all clicks. It's helpful
to distinguish these two cases -- for example, if a user scrolls past the top
results and clicks on the 10th one, then there's a clear ranking problem. But if
they click on the 1st, 2nd, 3rd, then 10th results, this is actually pretty good.

It also adds a label for the number of search results. That way, we can exclude
cases when there was a tiny number of results (< 3) and the user clicked on one
of these. This doesn't indicate a ranking success.

## Test plan

Expand UI unit test for logging clicks. I wasn't sure if/ how I could test the
metrics export locally 🤔

jtibshirani commented:
The dashboard I have in mind would show something like this: "the proportion of first clicks above position 5, for all searches with more than 5 results".

I'd like to play around with the data for a bit before uploading a new dashboard. Depending on how it looks, maybe we'd want to remove/ overwrite the old metric that tracks all clicks.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | -0.00% (-0.63 kb) | -0.01% (-0.63 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/jtibshirani/track-clicks/compare-report.html;tab=live_object) for a full comparison between the commits efed0f1 and 96d3cb0 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

limitedmage commented on client/branded/src/search-ui/results/StreamingSearchResultsList.tsx:
```diff
@@ -103,6 +102,7 @@ export const StreamingSearchResultsList: React.FunctionComponent<
     const { itemsToShow, handleBottomHit } = useItemsToShow(executedQuery, resultsNumber)
     const location = useLocation()
     const [rootRef, setRootRef] = useState<HTMLElement | null>(null)
+    const firstClick = useRef(true)
```
I think if the user clicks a result (not middle-click), then clicks the back button, and then clicks another result, `firstClick` will be true again because this component will be recreated. If that is the case but undesired, maybe we can move this logic to [`SearchResultCacheProvider`/`useCachedSearchResults`](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=main&file=client%2Fweb%2Fsrc%2Fsearch%2Fresults%2FSearchResultsCacheProvider.tsx&editor=VSCode&version=2.2.14&start_row=0&start_col=0&end_row=0&end_col=0) which caches the latest search so users can click the back button and not have everything reload again.

jtibshirani replied:
This is a great point, it wasn't working before! I pushed a new approach that keeps track of whether the results have been clicked using the cached search results.

I didn't feel super confident about the change -- please let me know if it's hacky. (When we don't have a feature freeze deadline I'm going to take some time to get up to speed with React Hooks and our frontend code :))

stefanhengl reviewed (approved):
The backend code looks good to me!

jtibshirani commented:
@limitedmage @stefanhengl I am hoping to revive this change for our search ranking work this release. I made the following changes:
* Rebased on top of latest `main` and and force-pushed 
* Fixed a major error in the "first click" logic
* Started recording whether the `search-ranking` feature is enabled, so we can compare metrics across both cases

jtibshirani commented:
On second thought, I decided to close this and open a new PR!
